core/tests/ts-integration/tests/fees.test.ts

```diff
@@ -128,12 +130,65 @@ testFees('Test fees', () => {
             );
         }
 
-        await setInternalL1GasPrice(alice._providerL2(), undefined, true);
-
         console.log(`Full report: \n\n${reports.join('\n\n')}`);
     });
 
+    test('Test gas consumption under large L1 gas price', async () => {
+        if (process.env.CHAIN_STATE_KEEPER_L1_BATCH_COMMIT_DATA_GENERATOR_MODE === 'Validium') {
+            // We skip this test for Validium mode, since L1 gas price has little impact on the gasLimit in this mode.
+            return;
+        }
+
+        // In this test we check that the server works fine when the required gasLimit is over u32::MAX.
+        // Under normal server behavior, the maximal gas spent on pubdata is around 120kb * 2^20 gas/byte = ~120 * 10^9 gas.
+
+        // In this test we will set gas per pubdata byte to its maximum value, while publishing a large L1->L2 message.
+
+        const minimalL2GasPrice = ethers.BigNumber.from(process.env.CHAIN_STATE_KEEPER_MINIMAL_L2_GAS_PRICE!);
+
+        // We want the total gas limit to be over u32::MAX, so we need the gas per pubdata to be 50k.
+        //
+        // Note, that in case, any sort of overhead is present in the l2 fair gas price calculation, the final
+        // gas per pubdata may be lower than 50_000. Here we assume that it is not the case, but we'll double check
+        // that the gasLimit is indeed over u32::MAX, which is the most important tested property.
+        const requiredPubdataPrice = minimalL2GasPrice.mul(100_000);
+
+        await setInternalL1GasPrice(
+            alice._providerL2(),
+            requiredPubdataPrice.toString(),
+            requiredPubdataPrice.toString()
+        );
+
+        const l1Messenger = new ethers.Contract(zksync.utils.L1_MESSENGER_ADDRESS, zksync.utils.L1_MESSENGER, alice);
+
+        // Firstly, let's test a successful transaction.
+        const largeData = ethers.utils.randomBytes(90_000);
+        const tx = await l1Messenger.sendToL1(largeData);
+        expect(tx.gasLimit.gt(UINT32_MAX)).toBeTruthy();
+        const receipt = await tx.wait();
+        expect(receipt.gasUsed.gt(UINT32_MAX)).toBeTruthy();
+
+        // Secondly, let's test an unsuccessful transaction with large refund.
+
+        // This should not be enough for the execution
+        const gasToPass = receipt.gasUsed.sub(1);
+        const unsuccessfulTx = await l1Messenger.sendToL1(largeData, {
+            gasLimit: gasToPass
+        });
```

Comment:
I'm wondering if tx with only 1 gas less will really always fail. E.g. if first and second transactions are in the same L1 batch then I think second will be cheaper because of IO refunds, no? Here are some suggestions:
- explicitly wait for L1 batch with first tx to be sealed before sending second one
- maybe try increasing data size instead of decreasing gasUsed? Given pubdata price is much higher it should be more robust

Reply:
When we restart the test we use `CHAIN_STATE_KEEPER_TRANSACTION_SLOTS=1` so it is expected that each batch has exactly one tx

Reply:
But agree, increasing data size is more robust, I'll try that